fdbserver/CommitProxyServer.actor.cpp

```diff
@@ -1167,6 +1167,71 @@ void determineCommittedTransactions(CommitBatchContext* self) {
 	}
 }
 
+inline int64_t extractTenantIdFromKeyRef(StringRef s) {
+	if (s.size() < TenantAPI::PREFIX_SIZE) {
+		return TenantInfo::INVALID_TENANT;
+	}
+	// Parse mutation key to determine tenant prefix
+	StringRef prefix = s.substr(0, TenantAPI::PREFIX_SIZE);
+	return TenantAPI::prefixToId(prefix, EnforceValidTenantId::False);
+}
+
+int64_t extractTenantIdFromSingleKeyMutation(MutationRef m) {
+	ASSERT(!isSystemKey(m.param1));
+
+	// The first 8 bytes of the key of this OP is also an 8-byte number
+	if (m.type == MutationRef::SetVersionstampedKey && m.param1.size() >= 4 && parseVersionstampOffset(m.param1) < 8) {
+		return TenantInfo::INVALID_TENANT;
+	}
+
+	ASSERT(isSingleKeyMutation((MutationRef::Type)m.type));
+
+	return extractTenantIdFromKeyRef(m.param1);
+}
+
+// Return true if a single-key mutation is associated with a valid tenant id or a system key
+bool validTenantAccess(MutationRef m) {
+	if (isSystemKey(m.param1))
+		return true;
+
+	if (isSingleKeyMutation((MutationRef::Type)m.type)) {
+		auto tenantId = extractTenantIdFromSingleKeyMutation(m);
```

Comment:
@sfc-gh-xwang don't we have to ensure that TenantID is found in the TenantMap; otherwise, the routine may respond `true` for a SingleMutation type, with KeyString having > 8 bytes, even when no such Tenant with corresponding TenantId is known to the cluster?

Reply:
Seems like `checkTenantNoWait` routine does lookup TenantMap but it isn't invoked in this version of overloaded `validTenantAccess`?